Add test sourceset support to MRJAR plugin

rjernst opened the pull request:
Mrjar plugin supports main sourcesets automatically configured with newer java versions. This commit adds support for test sourcesets as well. The key difference between main and test is that test sourcesets also extend their main counterparts.

rjernst committed:
Add test sourceset support to MRJAR plugin

Mrjar plugin supports main sourcesets automatically configured with
newer java versions. This commit adds support for test sourcesets as
well. The key difference between main and test is that test sourcesets
also extend their main counterparts.

elasticsearchmachine commented:
Pinging @elastic/es-delivery (Team:Delivery)

rjernst committed:
spotless

mark-vieira commented on build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/MrjarPlugin.java:
```diff
@@ -60,46 +61,39 @@ public void apply(Project project) {
-        SourceSet sourceSet = javaExtension.getSourceSets().maybeCreate(sourcesetName);
-        for (String parentSourceSetName : parentSourceSets) {
-            GradleUtils.extendSourceSet(project, parentSourceSetName, sourcesetName);
+        List<Integer> testVersions = findSourceVersions(project, MRJAR_TEST_SOURCESET_PATTERN);
```
I don't think we need to do this twice. We can assume for every main source set, there is a corresponding test source set, and if it's empty that's fine the tasks will just get skipped with `NO_SOURCE`.

rjernst replied:
Ah ok, didn't think about NO_SOURCE.

mark-vieira commented on build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/MrjarPlugin.java:
```diff
@@ -60,46 +61,39 @@ public void apply(Project project) {
+            String sourcesetName = SourceSet.TEST_SOURCE_SET_NAME + javaVersion;
+            if (mainVersions.contains(javaVersion)) {
+                // tests extend both their parent test source sets, and all of main sourcesets up to that version
+                testSourceSets.add(SourceSet.TEST_SOURCE_SET_NAME + javaVersion);
```
Should this be:
```suggestion
                testSourceSets.add(SourceSet.MAIN_SOURCE_SET_NAME + javaVersion);
```

rjernst replied:
Yep, copy/paste issue

mark-vieira commented on build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/MrjarPlugin.java:
```diff
@@ -60,46 +61,39 @@ public void apply(Project project) {
         jarTask.configure(task -> {
-            task.into("META-INF/versions/" + javaVersion, copySpec -> copySpec.from(sourceSet.getOutput()));
+            for (var version : versions) {
+                SourceSet sourceSet = javaExtension.getSourceSets().maybeCreate(baseSourcesetName + version);
```
Why are we attempting to create a source set here? Shouldn't this already exist? We also should definitely not be registering source sets in a task configuration action.

rjernst replied:
Oops, copy/paste issue, I'll fix

mark-vieira reviewed (commented):
I feel like I'm missing the bit where we configure the test _task_. How are these source sets making it on to the testing classpath? Also, should we have individual test tasks for each source set? How do we plan on conditionally handling tests that are applicable only to a particular runtime version?

rjernst commented:
> I feel like I'm missing the bit where we configure the test task.

In my mind I was thinking this was automatic with sourceset creation. Makes sense though that they are independent. I'll add that.

rjernst committed:
address feedback

rjernst commented:
@mark-vieira I believe I addressed all your comments in [1b2d055](https://github.com/elastic/elasticsearch/pull/106571/commits/1b2d0558ff5337dffff37867bc226557debac9d5)

rjernst committed:
Merge branch 'main' into build/mrjar_test_sourcesets

mark-vieira commented on build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/MrjarPlugin.java:
```diff
@@ -132,6 +132,44 @@ private void addMrjarSourceset(
+            testTask.dependsOn(jarTask);
+
+            SourceSetContainer sourceSets = GradleUtils.getJavaSourceSets(project);
+            FileCollection mainRuntime = sourceSets.getByName(SourceSet.MAIN_SOURCE_SET_NAME + javaVersion).getOutput();
```
Don't we need to remove _all_ the parent main source sets here. So if this was `test23` we'd need to remove `main21`, `main22` and `main23` or else we'll hit jar hell.

rjernst replied:
Of course, you are right. I've fixed this to loop over the main sourcesets up to that point.

mark-vieira commented on build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/MrjarPlugin.java:
```diff
@@ -132,6 +132,44 @@ private void addMrjarSourceset(
+
+    private void createTestTask(Project project, SourceSet sourceSet, int javaVersion) {
+        var jarTask = project.getTasks().withType(Jar.class).named(JavaPlugin.JAR_TASK_NAME);
+        project.getTasks().register(JavaPlugin.TEST_TASK_NAME + javaVersion, Test.class).configure(testTask -> {
```
We'll want to have the `check` task depend on this so that they actually get run.

mark-vieira commented on build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/MrjarPlugin.java:
```diff
@@ -132,6 +132,44 @@ private void addMrjarSourceset(
+            testTask.setClasspath(testRuntime.minus(mainRuntime).plus(project.files(jarTask)));
+
+            testTask.getJavaLauncher().set(
+                javaToolchains.launcherFor(spec -> { spec.getLanguageVersion().set(JavaLanguageVersion.of(javaVersion)); })
```
nit: we can omit the curly braces in this lambda.

mark-vieira commented on build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/MrjarPlugin.java:
```diff
@@ -60,46 +60,33 @@ public void apply(Project project) {
         var jarTask = project.getTasks().withType(Jar.class).named(JavaPlugin.JAR_TASK_NAME);
         jarTask.configure(task -> {
-            task.into("META-INF/versions/" + javaVersion, copySpec -> copySpec.from(sourceSet.getOutput()));
+            for (var version : versions) {
```
Looking at this more, this feels a bit awkward. I feel like we should be doing this in `addSourceSet` instead.

rjernst committed:
spotless

rjernst committed:
round 2

rjernst committed:
spotless

rjernst committed:
cleanup

rjernst committed:
Merge branch 'main' into build/mrjar_test_sourcesets
